### ISSUE TITLE
User migration process fails on really large user databases

### ISSUE THREAD

**Issue by pippinsplugins** [AffiliateWP/AffiliateWP#2180]
When a site has a massively large user database, the batch processing for user migration can fail. On [this customers's site](https://secure.helpscout.net/conversation/351640512/57613/) it was failing because the `pre_fetch()`, which sets the existing affiliates' user IDs in a temporary file, failed. This then resulted in zero accounts every getting migrated.

One possible way to resolve this will be to not exclude existing accounts from the `get_users()` query and, instead, always loop over _all_ user accounts and then determining if the user has an affiliate account already. If the user does not have an account, one is created for it.

The downside to this approach is that it requires a lot more queries and can take much, much longer to process, but it will create a reliable process that works on any size of database.

**Comment by DrewAPicture** [comment]
The query – despite pulling down a lot of user IDs – shouldn't be capable of failing with only 60k users. We're pulling down _only_ the user IDs from the affiliates table and not even the full affiliate objects.

And it's not storing it in a file, it's storing in an option via the temporary data store.

> One possible way to resolve this will be to not exclude existing accounts from the get_users() query and, instead, always loop over all user accounts and then determining if the user has an affiliate account already. If the user does not have an account, one is created for it.
>
> The downside to this approach is that it requires a lot more queries and can take much, much longer to process, but it will create a reliable process that works on any size of database.

I'm not sure this is a great approach. If we're talking about a site that can't pull down 60k integers from a single column in a single table, I'm not sure I want to try to pull down the full list of users on _every_ iteration of the batch process either.

**Comment by DrewAPicture** [comment]
My guess is that the problem is with storing the giant array more than the query.

**Comment by pippinsplugins** [comment]
>>>I'm not sure I want to try to pull down the full list of users on every iteration of the batch process either.

We wouldn't. With my proposed approach we'd iterate over all users, 20-30 at a time, and create affiliate accounts for any that do not have one.

**Comment by DrewAPicture** [comment]
Yeah, my main issue with that approach is that it's both incredibly innefficient and we're showing misleading progress information because we don't actually _know_ how many total affiliates we're creating until the very end.

Thinking about this a little bit, another approach could be to simply pull a list of all of the affiliate user IDs, as well as a list of all user IDs then store only the difference – the user IDs that don't exist in the affiliates table.

Then when we process a step, we're effectively stepping through the ids we explicitly know aren't affiliates. We don't even have to do queries with `get_users()`, just literally step through the IDs array each `per_step` and `offset` at a time.

**Comment by pippinsplugins** [comment]
>>>then store only the difference
Ooh, I like that approach!
